modules/store/src/selector.ts

```diff
@@ -17,13 +17,32 @@ export type ComparatorFn = (a: any, b: any) => boolean;
 
 export type DefaultProjectorFn<T> = (...args: any[]) => T;
 
+type ProjectorStrictness = 'strict';
+
+type SelectorProjectorArgs<
+  StrictnessConfig extends ProjectorStrictness,
+  ProjectorArgs
+> = StrictnessConfig extends 'strict'
+  ? ProjectorArgs extends unknown[]
+    ? ProjectorArgs
+    : any[]
+  : any[];
+
+type SelectorProjectorFn<ProjectorFn> = ProjectorFn extends (
+  ...args: infer ProjectorArgs
+) => infer ProjectorResult
+  ? <Strictness extends ProjectorStrictness = 'strict'>(
+      ...args: SelectorProjectorArgs<Strictness, ProjectorArgs>
+    ) => ProjectorResult
+  : ProjectorFn;
```

Comment:
If I understand this correctly (and feel free to correct me), these types are added to be able to "opt-out" of the strictness by providing `any` as the generic (`projector<any>(...)`)?

Would we still need extra types if we would not support this? 
For example as it was implemented in #2319

> Since the projector is found on arbitrary objects, then without compiling the code and inspecting type information, you could only naively update any call of a function named projector referenced as a property from an object.

You're right. Initially I was thinking we could do this only for spec|test files, but that could possibly update incorrect expressions.
On that note, do we need to allow `any`? I'm of the opinion that we shouldn't, because we want to encourage best practices. If a user still wants to be able to pass incorrect types to `projector` we shouldn't make it that easy, they can always do `(selector.projector as any)(arg1, arg2)` as a workaround.

Reply:
The hope was that strictly typed projectors could be introduced without being a breaking change. But, perhaps this is a case where a major version will introduce a type change that will require dependents to manually react. Hopefully it is valuable enough to force a change and not time consuming enough for dependents to resolve.

Reply:
Oh, to answer the other question, yes, the added type complexity is to support passing in the `any` generic. That complexity can be removed if we think people could just cast the projector itself as `any`.

Reply:
If providing a migration isn't possible then I'm in favor of removing the complexity, and make this a breaking change.